lib/active_merchant/billing/gateways/adyen.rb

```diff
@@ -45,34 +45,45 @@ def authorize(money, payment, options={})
         add_invoice(post, money, options)
         add_payment(post, payment)
         add_extra_data(post, options)
-        add_shopper_interaction(post,payment,options)
+        add_shopper_interaction(post, payment, options)
         add_address(post, options)
         commit('authorise', post)
       end
 
       def capture(money, authorization, options={})
         post = init_post(options)
-        add_invoice_for_modification(post, money, authorization, options)
-        add_references(post, authorization, options)
+        add_invoice_for_modification(post, money, options)
+        add_reference(post, authorization, options)
         commit('capture', post)
       end
 
       def refund(money, authorization, options={})
         post = init_post(options)
-        add_invoice_for_modification(post, money, authorization, options)
-        add_references(post, authorization, options)
+        add_invoice_for_modification(post, money, options)
+        add_original_reference(post, authorization, options)
         commit('refund', post)
       end
 
       def void(authorization, options={})
         post = init_post(options)
-        add_references(post, authorization, options)
+        add_reference(post, authorization, options)
         commit('cancel', post)
       end
 
+      def store(credit_card, options={})
+        requires!(options, :order_id)
+        post = init_post(options)
+        add_invoice(post, 0, options)
+        add_payment(post, credit_card)
+        add_extra_data(post, options)
+        add_recurring_contract(post, options)
```

Comment:
`add_recurring_contract` is called here an in `add_payment` is that intentional?

Reply:
Yes, it needs to be added both when a card is sent to be stored, and when a stored card is used for a transaction. No better way to add them in both situations is occurring to me right now but I'm open to suggestions.

Reply:
Fair enough. I don't have an alternative that would be any better. 👍